### PR TITLE
Renderer (excl animations)

### PR DESCRIPTION
**Pull request by Peter-JanGootzen**
# Renderer
This PR adds the renderer to the BrickEngine. It was made quite closely to the class diagram.

## What we learned during the creation of this PR
- Do not call make_unique on something you are just creating otherwise things will break, use the unique_ptr constructor instead.
- X11, Graphics drivers, SDL and many more libraries don't deallocate a ton of memory on exit. Ignore these when memory leak testing with valgrind.
- You can put a C struct with a destruction function inside a unique_ptr by using this as the type `std::unique_ptr<T, void(*)(T*)>` and then pass the destruction function along in the unique_ptr constructor.

## What we didn't do
- Animations, determining frames and working with texture sheets turned out to be quite hard and maybe our priorities should lie with other user stories first. But we did however design the renderer to work with animations, so we don't expect any architectural changes have to be made when animations are added.

### How to use this
Put these in assets/graphics/:
- https://drive.google.com/file/d/1jMgwvclmT8ugmw7b_LvcofIAFWggS26L/view?usp=sharing
- https://drive.google.com/file/d/1okMvVyoo4KMDIYdRQ8rn-yorREHS9RJ9/view?usp=sharing

```c++
#include <iostream>
#include <chrono>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <memory>

#include "brickengine/engine.hpp"
#include "brickengine/rendering/renderable_factory.hpp"
#include "brickengine/rendering/renderer.hpp"
#include "brickengine/rendering/renderables/renderable.hpp"
#include "class.hpp"

using namespace std::chrono_literals;

int main(int argc, char** argv) {
    BrickEngine engine = BrickEngine("Beast Arena", 800, 800, { 1, 2 });
    RenderableFactory* renderableFactory = engine.getRenderableFactory();
    auto dst_max = std::unique_ptr<Rect>(new Rect { 0, 0, 300, 300});
    std::unique_ptr<Renderable> makker_max = renderableFactory->createImage("./assets/graphics/verstappen_laughing.bmp", 1, std::move(dst_max));
    auto dst_vet = std::unique_ptr<Rect>(new Rect { 300, 150, 400, 600});
    std::unique_ptr<Renderable> makker_vet = renderableFactory->createImage("./assets/graphics/vettel_sochi.bmp", 2, std::move(dst_vet));
    Renderer* renderer = engine.getRenderer();

    int i = 0;
    while(true) {
        std::cout << "tick " << i << std::endl;
        std::this_thread::sleep_for(1s);
        makker_max->getDstRect()->x += i * 10;
        makker_max->getDstRect()->y += i * 10;
        renderer->clearScreen();
        renderer->queueRenderable(*makker_max.get());
        renderer->queueRenderable(*makker_vet.get());
        renderer->drawScreen();

        i++;
        if (i == 5)
            break;
    }

    return 0;
}
```

### CODE REVIEW

File: lib/engine.cpp [hunk 1]
```diff
@@ -1,69 +1,70 @@
...
-BrickEngine::BrickEngine() {
-    window = nullptr;
-    renderer = nullptr;
-    screenSurface = nullptr;
+void destory_window(SDL_Window* w) {
+    SDL_DestroyWindow(w);
 }
-BrickEngine::~BrickEngine() {
-
+void destory_renderer(SDL_Renderer* r) {
```
Comment (Bram-Boris) [inline lib/engine.cpp]: Typo

File: lib/engine.cpp [hunk 2]
```diff
@@ -1,69 +1,70 @@
...
+    this->renderer = std::shared_ptr<Renderer>(new Renderer(std::move(renderer_ptr), layers));
 
-    renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_SOFTWARE); 
- //    SDL_SetRenderDrawColor(renderer, 0, 0, 0, SDL_ALPHA_OPAQUE);
-//    SDL_RenderClear(renderer);
-//    SDL_RenderPresent(renderer);
+    this->renderableFactory = std::unique_ptr<RenderableFactory>(new RenderableFactory(renderer));
 
-    std::cout << "Window openend finsihed";
+    std::cout << "Window openend finsihed" << std::endl << std::endl;
```
Comment (Bram-Boris) [inline lib/engine.cpp]: Engrish? :P

File: include/brickengine/rendering/renderer.hpp [hunk 3]
```diff
@@ -2,16 +2,27 @@
...
-    void ClearScreen();
-    void DrawScreen();
-    void Render(Renderable renderable);
-    void Render(Image renderable);
-    void Render();
+    Renderer(std::unique_ptr<SDL_Renderer, void(*)(SDL_Renderer*)> r, std::vector<int> layers);
+    void clearScreen();
+    void drawScreen();
+    void queueRenderable(Renderable& r);
+    SDL_Texture* CreateTextureFromSurface(SDL_Surface* surface) const;
```
Comment (JanSchollaert) [inline include/brickengine/rendering/renderer.hpp]: CreateTextureFromSurface -> createTextureFromSurface?

**Comment by JanSchollaert** [comment]
Very nice :+1:

**Pull request merged**